ProjectTemplates/templates/Functions-CSharp/SampleFunc.cs

```diff
@@ -0,0 +1,127 @@
+                await req.HttpContext.AuthenticateAzureFunctionAsync();
+            if (!authenticationStatus) return authenticationResponse;
+
+            req.HttpContext.VerifyUserHasAnyAcceptedScope(scopeRequiredByApi);
+            using var response = await _downstreamWebApi.CallWebApiForUserAsync("DownstreamApi").ConfigureAwait(false);
+
+            if (response.StatusCode == System.Net.HttpStatusCode.OK)
+            {
+                var apiResult = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                // Do something with apiResult
+            }
+            else
+            {
+                var error = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                throw new HttpRequestException($"Invalid status code in the HttpResponseMessage: {response.StatusCode}: {error}");
+            }
+
+            string name = req.HttpContext.User.Identity.IsAuthenticated ? req.HttpContext.User.GetDisplayName() : null;
+
+            string responseMessage = string.IsNullOrEmpty(name)
+                ? "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response."
+                : $"Hello, {name}. This HTTP triggered function executed successfully.";
+
+            return new JsonResult(responseMessage);
+        }
+
+#elseif (GenerateGraph)
+        private readonly GraphServiceClient _graphServiceClient;
+
+        public SampleFunc(ILogger<SampleFunc> logger,
+            GraphServiceClient graphServiceClient)
+        {
+            _graphServiceClient = graphServiceClient;
+            _logger = logger;
+        }
+
+        [FunctionName("SampleFunc")]
+        public async Task<IActionResult> Run(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "get", "post", Route = null)] HttpRequest req)
+        {
+            _logger.LogInformation("C# HTTP trigger function processed a request.");
+
+            var (authenticationStatus, authenticationResponse) =
+                await req.HttpContext.AuthenticateAzureFunctionAsync();
+            if (!authenticationStatus) return authenticationResponse;
+
+            req.HttpContext.VerifyUserHasAnyAcceptedScope(scopeRequiredByApi);
+
+            var user = await _graphServiceClient.Me.Request().GetAsync();
+
+            string responseMessage = string.IsNullOrEmpty(user.DisplayName)
+                ? "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response."
+                : $"Hello, {user.DisplayName}. This HTTP triggered function executed successfully.";
+
+            return new JsonResult(responseMessage);
+        }
+#else
+        public SampleFunc(ILogger<SampleFunc> logger)
+        {
+            _logger = logger;
+        }
+
+        [FunctionName("SampleFunc")]
+        public async Task<IActionResult> Run(
```

Comment:
This method only needs to be async if `await req.HttpContext.AuthenticateAzureFunctionAsync();` is called (see below), that is `#if !NoAuth`

Do we want to have another big case

#elseif (!NoAuth)
// we would verify the token and the scopes, and return a Task<IActionResult>

#else
// no  verification of token and scope, return just a IActionResult

#endif

Reply:
fixed.

---
In reply to: [570245685](https://github.com/AzureAD/microsoft-identity-web/pull/903#discussion_r570245685) [](ancestors = 570245685)